Unroute source pin and preserve rest of net

clavin-xlnx opened the pull request:
This method is a complimentary method to `DesignTools.unroutePins()` in that it can unroute a source pin from a net while also keeping the routing from other source pins on the net intact.  This also sets the set of driven `SitePinInst` sinks from the unrouted routing to unrouted.

clavin-xlnx commented on src/com/xilinx/rapidwright/design/DesignTools.java:
```diff
@@ -1032,20 +1032,72 @@ public static boolean removeConnectedRouting(Net net, Node node){
 	/**
 	 * For the given set of pins, if they were removed, determine which PIPs could be trimmed as 
-	 * they no longer route to any specific sink.
+	 * they no longer route to any specific sink.  This method is intended only for sink pins.  
```
```suggestion
	 * they no longer route to any specific sink.  This method only works for sink pins.  
```

eddieh-xlnx reviewed (commented):
Overall comment: would it better to have `DesignTools.unroutePins()` handle both source and sink cases internally? This is the case for `Net.removePin()` which unroutes the whole net if it is a source pin, and for sink pins, either the whole net or just its branch (depending on the `preserveOtherRoutes` parameter)

eddieh-xlnx commented on src/com/xilinx/rapidwright/design/DesignTools.java:
```diff
@@ -1028,24 +1028,77 @@ public static boolean removeConnectedRouting(Net net, Node node){
+	/**
+	 * Unroutes a SitePinInst of a net.  This is desirable when a net has multiple SitePinInst
+	 * source pins (multiple outputs of a Site) and only a particular branch is desired to be 
+	 * unrouted.  If the entire net is to be unrouted, a more efficient method is {@link Net#unroute()}. 
```
Would it be wise to do this automatically -- if the net only has one source pin (assuming the alternate source is populated correctly) then it can go ahead and unroute the whole net?

clavin-xlnx replied:
In theory, yes we should.  But I'm wondering if we can always depend on the alternate source being populated correctly?  I guess we can make the change and find out, but perhaps we make that another PR.

clavin-xlnx commented:
> Overall comment: would it better to have `DesignTools.unroutePins()` handle both source and sink cases internally? This is the case for `Net.removePin()` which unroutes the whole net if it is a source pin, and for sink pins, either the whole net or just its branch (depending on the `preserveOtherRoutes` parameter)

Indeed, it would be better.  That was my original approach and it got messy because of the unique aspects of the source-removal case.  It wasn't working so I refactored and fixed the code to the current state.  Now that it is working, perhaps you could help me find a more elegant way to merge them.

eddieh-xlnx commented:
Does this handle routing stubs?
Here's a VU440 route that has one source pin, two sink pins, and a stub:
```
INT_X72Y538/INT.LOGIC_OUTS_E12->>INT_NODE_SINGLE_DOUBLE_76_INT_OUT
INT_X72Y538/INT.INT_NODE_SINGLE_DOUBLE_76_INT_OUT->>SS2_E_BEG3
INT_X72Y536/INT.SS2_E_END3->>INT_NODE_IMUX_71_INT_OUT
INT_X72Y536/INT.SS2_E_END3->>INT_NODE_SINGLE_DOUBLE_109_INT_OUT
INT_X72Y536/INT.INT_NODE_SINGLE_DOUBLE_109_INT_OUT->>WW2_E_BEG4
INT_X71Y536/INT.WW2_E_END4->>INT_NODE_SINGLE_DOUBLE_47_INT_OUT
INT_X71Y536/INT.INT_NODE_SINGLE_DOUBLE_47_INT_OUT->>NN2_E_BEG4
INT_X71Y538/INT.NN2_E_END4->>INT_NODE_SINGLE_DOUBLE_1_INT_OUT
INT_X71Y538/INT.INT_NODE_SINGLE_DOUBLE_1_INT_OUT->>EE2_E_BEG5
INT_X72Y538/INT.EE2_E_END5->>INT_NODE_IMUX_16_INT_OUT
INT_X72Y538/INT.INT_NODE_IMUX_16_INT_OUT->>BOUNCE_E_14_FTN
INT_X72Y538/INT.INT_NODE_IMUX_16_INT_OUT->>BYPASS_E13
```

clavin-xlnx commented:
Since I've tested this code and know that it works, it would be good to get it merged in as a reference point.  But, you make two good points:

1) Merge this functionality into `DesignTools.unroutePins()` and
2) Unroute the whole net if there is only one source pin.

How about we merge this and then I file an issue to resolve the other two issues?

eddieh-xlnx commented on src/com/xilinx/rapidwright/design/DesignTools.java:
```diff
@@ -1028,24 +1028,77 @@ public static boolean removeConnectedRouting(Net net, Node node){
+	        }
+	    }
+	    
+	    src.setRouted(false);
+        removePIPsFromNet(src.getNet(), pipsToRemove);
```
```suggestion
	    src.setRouted(false);
	    removePIPsFromNet(src.getNet(), pipsToRemove);
```

eddieh-xlnx commented on test/src/com/xilinx/rapidwright/design/TestDesignTools.java:
```diff
@@ -240,4 +237,115 @@ public void testGetTrimmablePIPsFromPins(String pinName) {
+    public void testRemoveSourcePin() {
+        Design design = new Design("test", Device.KCU105);
+        
+        Net net1 = createTestNet(design, "net1", new String[]{
```
```suggestion
        // Net with one source (AQ2) and two sinks (A_I & FX) and a stub (INT_NODE_IMUX_71_INT_OUT)
        Net net1 = createTestNet(design, "net1", new String[]{
```

eddieh-xlnx commented on test/src/com/xilinx/rapidwright/design/TestDesignTools.java:
```diff
@@ -240,4 +237,115 @@ public void testGetTrimmablePIPsFromPins(String pinName) {
+        }
+
+        
+        Net net2 = createTestNet(design, "net2", new String[]{
```
```suggestion
        // Net with one output (HMUX) and one input (SRST_B2)
        Net net2 = createTestNet(design, "net2", new String[]{
```
??

eddieh-xlnx commented on test/src/com/xilinx/rapidwright/design/TestDesignTools.java:
```diff
@@ -240,4 +237,115 @@ public void testGetTrimmablePIPsFromPins(String pinName) {
+        design.removeSiteInst(design.getSiteInstFromSiteName("SLICE_X64Y158"));
+        
+        
+        Net net3 = createTestNet(design, "net3", new String[]{
```
```suggestion
        // Net with two outputs (HMUX primary and H_O alternate) and two sinks (SRST_B2 & B2)
        Net net3 = createTestNet(design, "net3", new String[]{
```
???

eddieh-xlnx commented on test/src/com/xilinx/rapidwright/design/TestDesignTools.java:
```diff
@@ -240,4 +237,115 @@ public void testGetTrimmablePIPsFromPins(String pinName) {
+            SitePinInst altSnk = net3.createPin("B2", si);
+            altSnk.setRouted(true);
+            
+            Set<PIP> unroutedPIPs = DesignTools.unrouteSourcePin(net3.getAlternateSource());
```
```suggestion
            // Unroute just the H_O alternate source
            Set<PIP> unroutedPIPs = DesignTools.unrouteSourcePin(net3.getAlternateSource());
```
